File: src/github.com/xlab/cgogen/generator/test/proxy_test.go
package test

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTestProxyString(t *testing.T) {
	expected := "HELLO"
	result := TestProxyString("a", "b", "c")
	assert.Equal(t, expected, result)
}

func TestTestProxyBytes(t *testing.T) {
	expected := []byte("abcd")
	result := TestProxyBytes([]byte("ab"), []byte("cd"))
	assert.Equal(t, expected, result)
}

func TestTestProxyPtr3String(t *testing.T) {
	cube := [4][][][]string{
		[][][]string{
			[][]string{
				[]string{"a1", "b1"},
				[]string{"c1", "d1"},
				[]string{"i1", "j1"},
				[]string{"k1", "l1"},
			},
			[][]string{
				[]string{"a2", "b2"},
				[]string{"c2", "d2"},
				[]string{"i2", "j2"},
				[]string{"k2", "l2"},
			},
			[][]string{
				[]string{"a3", "b3"},
				[]string{"c3", "d3"},
				[]string{"i3", "j3"},
				[]string{"k3", "l3"},
			},
		},
		[][][]string{
			[][]string{
				[]string{"a1", "b1"},
				[]string{"c1", "d1"},
				[]string{"i1", "j1"},
				[]string{"k1", "l1"},
			},
			[][]string{
				[]string{"a2", "b2"},
				[]string{"c2", "d2"},
				[]string{"i2", "j2"},
				[]string{"k2", "l2"},
			},
			[][]string{
				[]string{"a3", "b3"},
				[]string{"c3", "d3"},
				[]string{"i3", "j3"},
				[]string{"k3", "l3"},
			},
		},
	}

	result := TestProxyPtr3String(cube)
	assert.Equal(t, [][][]string{}, result)
}

File: src/github.com/xlab/cgogen/generator/test/proxy_test.c
#include "proxy_test.h"

const char *test_proxy_string(const char *s1, const char *s2, const char *s3) {
	if (strcmp(s1, "a") == 0 && 
		strcmp(s2, "b") == 0 && 
		strcmp(s3, "c") == 0) {
		 	return "HELLO";
		}
	return "BYE";
}

unsigned char* test_proxy_bytes(unsigned char *buf, char *buf_two) {
	unsigned char* result = (unsigned char*)malloc(4);
	memcpy(result, buf, 2);
	memcpy(&result[2], buf_two, 2);
	return result;
}

const char *****test_proxy_ptr3_string(const char ****str[4]) {
	return str;
}
